Support environment variables from a directory

mattrobenolt opened the pull request:
This adds two new arguments, `--env-dir` and `--label-dir` and are modeled after `envdir`. See: http://cr.yp.to/daemontools/envdir.html

```
# ls -lh /env
total 12K
-rw-r--r-- 1 root root 4 May 14 22:34 foo
-rw-r--r-- 1 root root 4 May 14 22:42 foo2
-rw-r--r-- 1 root root 7 May 14 22:54 nullbyte

# cat /env/*
bar
bar
foobar

# /go/src/github.com/docker/docker/bundles/1.7.0-dev/binary/docker-1.7.0-dev run --env-dir /env ubuntu:14.04 env
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HOSTNAME=0241be89c367
foo=bar
foo2=bar
nullbyte=foo
bar
HOME=/root
```

mattrobenolt committed:
Support environment variables from a directory

Signed-off-by: Matt Robenolt <matt@ydekproductions.com>

GordonTheTurtle commented:
Can you please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "env-dir" git@github.com:mattrobenolt/docker.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

This will update the existing PR, so **DO NOT** open a new one.

duglin commented:
@mattrobenolt just curious, do other tools support things like this?  I've never seen this type of thing before.

mattrobenolt commented:
Yeah, the tool I linked to `envdir` utilizes this pattern: http://cr.yp.to/daemontools/envdir.html
`runit` also has this built in as a part of it's suite of tools: http://smarden.org/runit/chpst.8.html (See the `-e dir` option)
`mesos` as well utilizes a similar pattern for defining `attributes` on a node. You can use a directory of files in this exact manner.

mattrobenolt commented:
To add to it, it's also easier to manage from config management systems, since different pieces can just drop in a file without coordinating between each other. In our case, we're populating machines with a bunch of env variables into `/etc/env.d`, and would like to then just link up `--env-dir /etc/env.d` and see magic happen. :sparkles:

crosbymichael commented:
It's a pretty interesting idea and I have used this with `mesos` and it works really well for automation.  I guess my question is that if we add this do you see demand for other things out of `env`?  Is this useful only for `env` or can it be used other places?

mattrobenolt commented:
@crosbymichael I'm not entirely sure of all of the options that would be suitable with docker. I'm a pretty light docker user myself, but this patch also covers labels, just because they utilize the exact same underlying code, so `--label-dir` is being exposed as well.

mattrobenolt commented:
Anything else I can do here to help get this pushed through? This feature is becoming increasingly more useful for us, and I'd hate to run a fork in production. :(

thaJeztah commented on opts/envdir.go:
```diff
@@ -0,0 +1,43 @@
+		// ignore directories
+		if !info.IsDir() {
+			variable := info.Name()
+			if strings.ContainsAny(variable, whiteSpaces) {
```
Given that this directly uses the filename as the name of the env-var, I can think of other file names to excluded as well (hidden files?) also, only names that are valid for environment variables?

A quick search for what characters are allowed; http://stackoverflow.com/questions/2821043/allowed-characters-in-linux-environment-variable-names

thaJeztah replied:
^^ not really relevant in this stage, so just ignore for now

mattrobenolt replied:
So I mostly agree, but I feel like doing that is out of scope of this function as well as out of scope for this PR.

If I were to implement more strict validation, I'd do it later after the collection of all the values that way even CLI args passed in through `-e FOO=bar` style were all validated the same way.

The role of `ParseEnvDir` should do it's one thing, and that's crawl a directory and convert into `k=v` pairs, which get passed along later down the pipeline.

mattrobenolt replied:
Also worth noting that when I originally wrote `ParseEnvDir`, I skipped "hidden" files, that began with `.`, but opted out of it because `envdir` does not skip them. Maybe there's valid reason to name a var `.FOO`? I've never done it, but I don't see rejecting the idea.

And to my previous point, we don't disallow this inside of `-e .FOO=bar` or in an `env-file`.

mattrobenolt replied:
I just checked out the original `envdir.c` and it does, in fact, ignore files that begin with `.`. Will update to reflect this.

For posterity, here is the actual `envdir.c`: https://gist.github.com/mattrobenolt/1e7621accfe2377dde0c

thaJeztah commented:
@crosbymichael @calavera @duglin what are your thoughts on this? Is this something that's wanted? It looks fairly simple.

I wonder if there should be restrictions on the paths that can be used for the env-dir; should it be within the build context?

duglin commented:
I do like the fact that the amount of code needed to support it is minimal.  Would need some testcases and docs too, if we go this route.
I'm just not familiar enough with this way of specifying env vars to have much of an opinion on whether this is something popular enough to add, or if its just a passing fad.  If I was forced to choose, I'd probably let it in since it doesn't break or complicate anything in Docker, it doesn't clutter things up, and if people don't like it then they don't need to use it.

mattrobenolt commented:
> Would need some testcases and docs too, if we go this route.

I have no issues helping in this area. Not entire sure where documentation is held atm, and none of this environment variable parsing stuff has tests atm. (also not super familiar with testing in Go, let alone for docker), but absolutely willing to contribute.

mattrobenolt commented:
> I wonder if there should be restrictions on the paths that can be used for the env-dir; should it be within the build context?

Not sure what all this entails, but we do this in say, `/etc/env.d` atm on our machines. Not sure what exactly the "build context" covers.

> Also, might need some guards, pointing it to a directory with huge files in it would be interesting ;)

Is this different than pointing it at a 10MB `env-file`? Aside from all the IO of reading all of the files. I'm ok imposing an artificial limit, but not sure what is reasonable in that case. I wouldn't feel too comfortable with anything less than 1000. At least we won't recursively crawl directories and subdirectories. ;) That at least prevents something really bad from happening if you were to accidently do `-env-dir=/`.

cpuguy83 commented:
I definitely don't like the extra flags. Seems like `--env-file` can be used to import a dir as well.

mattrobenolt commented:
@cpuguy83 It can be, but I personally feel that seems a bit misleading IMO since it's literally named `env-file`. But I'm not super strongly opposed if others agree with that. I'll throw in my -1 though. :)

mattrobenolt commented:
I've updated to ignore hidden files in the `env-dir` directory.

icecrime commented:
Thanks @mattrobenolt, that's an interesting feature :-)

+1 on design, even with an additional command-line flag (I hate adding flags, but I hate ambiguous inputs even more).

Pushing to code review.

icecrime commented on opts/envdir.go:
```diff
@@ -0,0 +1,43 @@
+				return []string{}, ErrBadEnvVariable{fmt.Sprintf("variable '%s' has white spaces", variable)}
+			}
+
+			data, err := ioutil.ReadFile(filepath.Join(directory, variable))
```
Disagree with `ReadFile` here: the [`envdir`](http://cr.yp.to/daemontools/envdir.html) page mentions it only reads the **first-line**:

> If  d contains a file named s whose **first line** is t [...]

icecrime commented:
One issue with implementation. Also this absolutely needs some tests: either unit tests through the regular `go` testing mechanism (`*_test.go` files basically), or look into `integration-cli/` subdirectory for examples of integration tests.

mattrobenolt commented:
@icecrime Thanks for the feedback. I'll happily write some tests for it. Just wanted to make sure that the idea was acceptable first before committing to that + docs.

mattrobenolt commented:
Also, just double checked, and envdir does in fact only read the first line. Updating behavior today + tests.

mattrobenolt commented:
@icecrime I've added some tests explicitly for the new `ParseEnvDir` function. I was going to add tests for the `ParseEnvFile` path as well, but see #13694.

Not sure why the "janky" build has failed, but tests passed locally for me and from what I can tell, shouldn't be related to what I've added.

thaJeztah commented:
Kicked janky to try again. Fingers crossed :-)

mattrobenolt commented:
@thaJeztah Looks to have failed again for a different unrelated issue. :disappointed:

thaJeztah commented:
Let's try once more (kicked). I'm on my mobile now so didn't check which test failed this time.

duglin commented on opts/envdir.go:
```diff
@@ -0,0 +1,56 @@
+		variable := info.Name()
+		// ignore directories and hidden files
+		if !info.IsDir() && variable[0] != '.' {
+			if strings.ContainsAny(variable, whiteSpaces) {
```
I think we need to check for more than just spaces and tabs:
Variables set by the user must have a name consisting solely of alphabetics, numerics, and underscores - the first of which must not be numeric.

mattrobenolt replied:
> I think we need to check for more than just spaces and tabs:
Variables set by the user must have a name consisting solely of alphabetics, numerics, and underscores - the first of which must not be numeric.

If we want to be more restrictive, that's fine, but I'd suggest we do this elsewhere and cover all the cases of setting environment variables in Docker, not just in envdir.

fwiw, I don't think that's a requirement:
```
$ docker run --rm -e '~0SERVICE_NA-M E$&!@#=foo' -it ubuntu:14.04 env | grep SERVICE
~0SERVICE_NA-M E$&!@#=foo
```

duglin replied:
LOL trying to set it from bash/sh fails for me. 
Agreed we should have a generic "check env var name" func some place

mattrobenolt replied:
Yeah, not entirely sure what all of the restrictions there are, or if we should care. But as is, I don't want to address this with this pull request. Is that fair?

vdemeester replied:
Updated #13694 with the regexp check. I've create an proposal issue https://github.com/docker/docker/issues/13824 (I definitly should update the content though) to list the checks that we do several times (like check env var & co) to centralize them a bit.

mattrobenolt replied:
Oh, neat. I can wait until #13694 lands in master, then rebase against that and use that regexp.

duglin replied:
Are we going to wait for #13694?

mattrobenolt replied:
yay, I've rebased and updated this to utilize #13694's new pattern.

duglin commented on opts/envdir.go:
```diff
@@ -0,0 +1,56 @@
+				"%s=%s",
+				variable,
+				// null bytes are to be converted to newlines according to envdir
+				bytes.TrimSpace(bytes.Replace(data, []byte{0}, []byte{'\n'}, -1)),
```
I'd prefer if you did this clean-up logic in `getFirstLine` just to keep this call to append() a little less complex.

duglin replied:
Did you want to handle the case of the file being empty? per the doc you reference it should remove the env var, right?

duglin replied:
per the doc you referenced, \0 is replace with \n but that seems kind of weird and not very user-friendly.
Is that doc a standard or just one person's view of how this should be done?
It just seems to me that rather than converting \0 to \n and ONLY using the first line of the file, why not read in the entire file 'as is' and let the user put line-breaks in there instead of \0's.

mattrobenolt replied:
> I'd prefer if you did this clean-up logic in getFirstLine just to keep this call to append() a little less complex.

Will do.

> Did you want to handle the case of the file being empty? per the doc you reference it should remove the env var, right?

I'd like to, but I think that's out of scope for this pull request since that'd require a lot more refactoring and thought. For example, how does that interop with `env-file` and just `e`? And nothing down the pipeline allows removing env variables from the host container. I also think this is less useful in this context since a container by nature is more isolated.

> per the doc you referenced, \0 is replace with \n but that seems kind of weird and not very user-friendly.
Is that doc a standard or just one person's view of how this should be done?
It just seems to me that rather than converting \0 to \n and ONLY using the first line of the file, why not read in the entire file 'as is' and let the user put line-breaks in there instead of \0's.

Kinda weird, I agree. I'd like to conform to what envdir implements so that a directory can interop and tooling that's designed to generate envdir compatible things continue to work. This behavior has been adopted by other implementations of envdir as well. See: https://github.com/jezdez/envdir/blob/master/envdir/env.py#L66

duglin commented on opts/envdir_test.go:
```diff
@@ -0,0 +1,65 @@
+			}
+		} else {
+			if err != nil {
+				t.Fatal(err)
```
can we get the 'dir' (or something) printed so we know which test we were on when it failed?  Just in case 'err' doesn't contain anything to tell us.

duglin commented on opts/fixtures/envdir/junk/.FOO:
```diff
@@ -0,0 +1 @@
+bar
```
change this to something other than 'bar' otherwise we won't know if it picked it up since it will overlap with the "FOO" file's value.

duglin commented on opts/fixtures/envdir/trim/FOO:
```diff
@@ -0,0 +1 @@
+  bar   
```
Since we ignore 2+ lines, go ahead and add a comment saying that there are spaces after 'bar' on the first line so that people don't erase them by mistake.

duglin commented on opts/envdir_test.go:
```diff
@@ -0,0 +1,65 @@
+		{"fixtures/envdir/trim", []string{"FOO=bar"}, false},
+		{"fixtures/envdir/null", []string{"FOO=ba\nr"}, false},
+		{"fixtures/envdir/empty", []string{}, false},
+		{"fixtures/envdir/whitespace", []string{}, true},
```
did you forget to add the "whitespace" dir?
If you meant for it to test a missing dir, use a better name like "nodir".
Then add a 'whitespace' dir to test bad file names.

mattrobenolt replied:
Oops, forgot to add. :)

duglin commented:
I think we need some docs and, in particular, a statement about the order in which things will be processed? Which env var setting wins.

mattrobenolt commented:
@duglin Feedback should be addressed. Not sure why tests are failing on the `test-docker-py` part, but they work for me locally. Seems highly unlikely to be related, but I'm open to ideas. :)

mattrobenolt commented:
@icecrime Anything else to be done here? Seems we have stalled a bit. I was last waiting on #13694 since that's affecting similar code, but that seems to have been stalled for the past 5 days. Would like to push this forward since this is important to us. :)

icecrime commented:
Thanks, LGTM :+1: Ping @duglin!

duglin commented:
@mattrobenolt did you want to wait for #13694?

mattrobenolt commented:
@duglin It's not my call. :) I would have liked to wait for the rest of tests for this module, but nothing in this PR directly affects it, so it doesn't particularly matter to me.

tiborvass commented:
Am I the only one who really doesn't want two new flags for little added value? :(
What's the usecase? If the usecase is "why not?" then I'd rather not add more features. People can just as well use `--env-file` and `--label-file`.

mattrobenolt commented:
> People can just as well use --env-file and --label-file

These are not replacements for my proposal. What I'm proposing is a very different semantic. And `env-file` is a file with a bunch of environment variables in it split by newlines. And `env-dir` is a directory, where each file is it's own variable, and the contents of the file is the value of the variable.

> for little added value?

Isn't that a pretty subjective assumption? It's very valuable for me and our use case. :) As stated, this is a pre-existing pattern used by runit as well, so this isn't a new invention.

cpuguy83 commented:
I totally agree with @tiborvass 
In any case, docker could detect if the path you are specifying is a dir or a file and parse accordingly.

mattrobenolt commented:
> docker could detect if the path you are specifying is a dir or a file and parse accordingly.

I'm ok with that.

Do you think it'd be weird to have two different behaviors based on if the flag is pointed at a dir or at a file? This is my only objection, but it's not a strong one.

mattrobenolt commented:
Oh, it seems this was already brought up here: https://github.com/docker/docker/pull/13234#issuecomment-105603646 :)

tiborvass commented:
@mattrobenolt but what's the usecase other than having another way of representing the same thing and that others do it too?

`for i in $dir*; do echo $(basename $i)=$(cat $i); done > $file` converts easily a directory into a file.

I would prefer not to overload `--env-file` because that is really weird, nobody does that and nobody expects it to work. My question is do we really need to support `--env-dir` or can't we let others have a wrapper in bashrc.

/cc @tianon

mattrobenolt commented:
I mean, at this point, I honestly don't care. It was a nice little feature that makes it nice for us to work with. I've explained my use case and cited examples of how this is used in the wild. For us, we utilize runit a lot, and runit has adopted the pattern of `envdir`s. There are many projects that have adopted envdir and different ports that provide the ability to parse envdirs.

For us, we dump out files from config management software which is much easier when you can manage single files, rather than one big file with everything in it.

We can't run `for i in $dir/*; do echo $(basename $i)=$(cat $i); done > $file` to do this nicely because our images are being run my mesos. So we pass the `--env-file` now directly. There's no way for this to be wrapped in a bash script.

If there's such resistance, I'm ok closing the pull request.

tiborvass commented:
@mattrobenolt sorry if you felt my tone was offending, it really is not my intention. I'm just one voice, other maintainers disagree with me so you don't need to close your pull request :)
I'm simply trying to understand the problem and try to find ways to solve it without adding more *nice little features* to docker.

Anyway, if really adding `--env-dir` and `--label-dir` is the best solution to the issue at hand, then so be it.

mattrobenolt commented:
@tiborvass nah, it's fine. Just frustrating from a very avid open source contributor trying to contribute to a new project.

I also understand the burden of adding new things and needing to maintain that. I deal with it all the time for my own projects, so I respect that.

Just trying to find a middle ground that works within the project and satisfies our scenario. Trust me, I'd really rather not run a fork of docker and maintain that. ;) So I'm trying to do whatever I can to marshall this through.

tiborvass commented:
@mattrobenolt thanks, no worries :)

Also, is everyone okay with `--env-dir` having highest precedence? I'm now also wondering if `--env-file` having higher precedence than `--env` is a good thing... Intuitively I would have put *dir < file < var* but maybe it's just me.

icecrime commented:
I'm fine, and still LGTM. @cpuguy83 @tiborvass Let's try and finish this one!

tonistiigi commented:
from spec:
> If the file s is completely empty (0 bytes long), envdir removes an environment variable named s if one exists, without adding a new variable.

This case doesn't seem to be handled and actually errors with `EOF` in client atm.

There is also a rule for no `=` in filename. Not sure if this should error or trim though.

mattrobenolt commented:
> This case doesn't seem to be handled and actually errors with EOF in client atm.

I wasn't sure how to handle this case without a decent size refactor since as of now, all of the environment variables get collected and added up together into a list. It'd need to be refactored to change that to be able to remove items from that list while accumulating them.

> There is also a rule for no = in filename. Not sure if this should error or trim though.

I'll double check against envdir source and see what it does.

> Needs docs as well.

Would be happy to add after code review.

tonistiigi commented:
> I wasn't sure how to handle this case without a decent size refactor since as of now, all of the environment variables get collected and added up together into a list. It'd need to be refactored to change that to be able to remove items from that list while accumulating them.

It at least needs to clear the variable then and not `EOF`.

mattrobenolt commented:
I can clear the variable within the context of `ParseEnvDir`, but it won't be able to remove any other environment variables, so in my opinion, that'd be a bit misleading unless we were able to delete any environment variables. Also not even sure being able to arbitrarily delete environment variables like this is a good fit for docker. What happens if someone were to try and delete `HOSTNAME`, etc? I just think it's a bit more dangerous to do in docker than in runit. But I'm open to hearing opinions. Again, would also require a pretty large refactor likely and haven't looked into it. Right now, all variables that we add are additive and piggyback of the same internals. Would we want to allow people to delete vars through `-e`? It opens the door for lots of questions. :)

> and not `EOF`.

Totally reasonable. Will update.

tonistiigi commented:
Sorry. I meant set the variable to be empty. This is what you can do with current -e as well.

mattrobenolt commented:
Ah, alright. Updating.

mattrobenolt commented:
@tagomoris Updated. One last question though re:

> There is also a rule for no = in filename. Not sure if this should error or trim though.

In envdir.c, the file is ignored. In all of docker, this doesn't really break things, but it's also undesired behavior. This check doesn't exist for `env-file` nor passing through `-e`. If you do `-e FOO=bar=baz`, you end up with one variable named `FOO` value being `bar=baz`.

Is the sane default here to ignore the variable entirely for `env-dir`? That seems reasonable to me.

mattrobenolt commented:
@tonistiigi Alright, since I've rebased and switched to using `EnvironmentVariableRegexp`, this will not return a `ErrBadEnvVariable` if there's an `=` in the name. This follows in line with what `env-file` does now as well.

tonistiigi commented:
code LGTM

Would not hurt to add an `integration-cli` test so that we actually test the functionality. Not only the parser function.

duglin commented:
so far it looks good to me but as @tonistiigi said, let's add some integration tests to make sure the entire processing flow works end-to-end.

mattrobenolt commented:
Alright, lemme look into how to do that. :+1:

thaJeztah commented:
Thanks for being so patient @mattrobenolt, sorry it's taking so long :+1:

bfirsh commented:
Thanks for your patience @mattrobenolt! Any update on the integration tests?

tiborvass commented on opts/envdir.go:
```diff
@@ -0,0 +1,63 @@
+func ParseEnvDir(directory string) ([]string, error) {
+	dir, err := ioutil.ReadDir(directory)
+	if err != nil {
+		return []string{}, err
```
@mattrobenolt is there a reason you don't `return nil, err` ?

diogomonica commented on opts/fixtures/envdir/trim/FOO:
```diff
@@ -0,0 +1,2 @@
+  bar   
+// The spaces after "bar" are intention, do not remove
```
*intentional

diogomonica commented:
Quick comments:
- Overall LGTM (slight feature creep)
- It isn't obvious to me that right now the priority is dir < file < var in terms of which env variable wins when there are competing variables. I'm not super familiar with the code, but there should be a test enforcing this behavior.
- An integration test would be nice

jfrazelle commented:
I agree this is feature creeep and am still not convinced we need to support it :/

bfirsh commented:
FWIW I haven't heard of anybody asking for something like this. @mattrobenolt is there a specific reason why you can't use a single file?

mattrobenolt commented:
@bfirsh https://github.com/docker/docker/pull/13234#issuecomment-102208876

jfrazelle commented:
Hi @mattrobenolt ,

Thank you so much for taking the time to send this PR. And I am so sorry it has gone for so long and had so much back and forth. We are trying to make our process better and unfortunately this PR was opened when we are still in our "trying to be better about keeping things open for a long time" phase.
However, I do not think this feature is super necessary, but I do think if people really want it they should express their wishes for it here.
See my fear is this will become like `docker rename` something a lot of people wanted but unfortunnately no one used, so it took us 3 releases to realize it was broken. Since this PR lacks any integration testing, this possibility could totally come true. Again thank you so much for taking the time to submit this. And so sorry for the delay in feedback.
If it turns out a ton of people comment here about their wishes to have this feature, then I think we can reconsider.

mattrobenolt commented:
@jfrazelle I appreciate the feedback. :heart:

Fortunately, we already worked around the issue 4 months ago, and this was just a nice-to-have, but not a deal breaker.

I didn't expect a resounding "yes", was just hoping to get some answer before investing more time into writing docs/integration tests on if the feature was even acceptable in the project. :)

icecrime commented:
Thanks for the understanding @mattrobenolt :+1:
